### PR TITLE
Modeling fixes for heterogeneous devices

### PR DESCRIPTION
**Pull request by carmocca**
See added comments

### CODE REVIEW

File: generate/base.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ def sample(logits: torch.Tensor, temperature: float = 1.0, top_k: Optional[int]
 def next_token(model: GPT, input_pos: torch.Tensor, x: torch.Tensor, **kwargs: Any) -> torch.Tensor:
     logits = model(x, input_pos)
     next = sample(logits, **kwargs)
-    return next.type_as(x)
```
Comment (carmocca) [inline generate/base.py:52]: Turns out that `type_as` will also change the device. PyTorch has soft-deprecated it for this reason.

File: lit_gpt/model.py [hunk 2]
```diff
@@ -48,8 +48,10 @@ def max_seq_length(self, value: int) -> None:
             cos, sin = self.rope_cache()
             self.register_buffer("cos", cos, persistent=False)
             self.register_buffer("sin", sin, persistent=False)
+        # overrides
+        elif self.cos.device.type == "meta":
+            self.cos, self.sin = self.rope_cache()
```
Comment (carmocca) [inline lit_gpt/model.py:53]: This is so you can do

```python
with meta:
    # __init__ will create the rope cache, which is going to be on meta
    model = GPT()
with device:
    # recreate the rope cache on an actual device
    model.max_seq_length = model.max_seq_length
```

File: lit_gpt/model.py [hunk 3]
```diff
@@ -343,3 +349,8 @@ def forward(self, input_pos: torch.Tensor, k: torch.Tensor, v: torch.Tensor) ->
     def reset_parameters(self) -> None:
         torch.nn.init.zeros_(self.k)
         torch.nn.init.zeros_(self.v)
+
+
+def build_mask_cache(max_seq_length: int, device: Optional[torch.device] = None) -> torch.Tensor:
```
Comment (carmocca) [inline lit_gpt/model.py:348]: I need to recreate the mask cache externally, so I extracted this bit out
